File: packages/oxlint-plugin-react-doctor/src/plugin/rules/a11y/radix-dialog-content-requires-title.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getTrailingJsxNameSegment } from "../../utils/get-trailing-jsx-name-segment.js";
import { hasJsxPropIgnoreCase } from "../../utils/has-jsx-prop-ignore-case.js";
import { hasJsxSpreadAttribute } from "../../utils/has-jsx-spread-attribute.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { resolveNamespacedPartName } from "../../utils/resolve-namespaced-part-name.js";
import { resolveShadcnUiComponentName } from "../../utils/resolve-shadcn-ui-component-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { scanJsxSubtreeForPart } from "../../utils/scan-jsx-subtree-for-part.js";
import type { RuleVisitors } from "../../utils/rule-visitors.js";

const RADIX_UNIFIED_MODULE_PATTERN = /^radix-ui$/;

interface RadixDialogSurface {
  // "Dialog" / "AlertDialog" — the unified package's namespace export and
  // the human name in diagnostics.
  readonly namespaceName: string;
  // The per-primitive package ("@radix-ui/react-dialog"), whose exports are
  // the bare parts (Content, Title, …).
  readonly primitiveModulePattern: RegExp;
}

// Radix's Dialog primitive requires a Title inside Content — without one the
// dialog has no accessible name and Radix logs a runtime accessibility error.
const RADIX_DIALOG_SURFACES: ReadonlyArray<RadixDialogSurface> = [
  { namespaceName: "Dialog", primitiveModulePattern: /^@radix-ui\/react-dialog$/ },
  { namespaceName: "AlertDialog", primitiveModulePattern: /^@radix-ui\/react-alert-dialog$/ },
];

const NAME_PROVIDING_ATTRIBUTES = ["aria-label", "aria-labelledby", "title"] as const;

// Resolves an element to a Radix part name for one surface, across both
// import styles: `import * as Dialog from "@radix-ui/react-dialog"` /
// `import { Content } from "@radix-ui/react-dialog"` (per-primitive) and
// `import { Dialog } from "radix-ui"` with `<Dialog.Content>` (unified).
const resolveRadixPartName = (
  elementName: EsTreeNode,
  surface: RadixDialogSurface,
  context: RuleContext,
): string | null =>
  resolveShadcnUiComponentName(elementName, surface.primitiveModulePattern, context) ??
  resolveNamespacedPartName(
    elementName,
    RADIX_UNIFIED_MODULE_PATTERN,
    surface.namespaceName,
    context,
  );

const isTitleElementName = (
  elementName: EsTreeNode,
  surface: RadixDialogSurface,
  context: RuleContext,
): boolean => {
  if (resolveRadixPartName(elementName, surface, context) === "Title") return true;
  // A local wrapper named like the title part ("Title", "DialogTitle") —
  // trusting the name trades a rare false negative for zero noise.
  const trailingSegment = getTrailingJsxNameSegment(elementName);
  return trailingSegment === "Title" || trailingSegment === `${surface.namespaceName}Title`;
};

export const radixDialogContentRequiresTitle = defineRule({
  id: "radix-dialog-content-requires-title",
  title: "Radix dialog content without a title",
  severity: "warn",
  requires: ["radix-ui"],
  recommendation:
    "Give every Radix Dialog.Content and AlertDialog.Content a Title part (wrapped in VisuallyHidden when the design shows no heading) or name the dialog with aria-label.",
  create: (context: RuleContext): RuleVisitors => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      for (const surface of RADIX_DIALOG_SURFACES) {
        if (resolveRadixPartName(node.name, surface, context) !== "Content") continue;
        // A spread can supply aria-label / children at runtime.
        if (hasJsxSpreadAttribute(node.attributes)) return;
        if (
          NAME_PROVIDING_ATTRIBUTES.some((attribute) =>
            hasJsxPropIgnoreCase(node.attributes, attribute),
          )
        ) {
          return;
        }
        const element = node.parent;
        if (!element || !isNodeOfType(element, "JSXElement") || element.children.length === 0) {
          return;
        }
        const scan = scanJsxSubtreeForPart(element.children, {
          isPartElementName: (elementName) => isTitleElementName(elementName, surface, context),
          // Same-surface parts (Close, Description, Overlay, …) are known
          // leaves; any unresolved custom component may render the title
          // itself, so the claim becomes unprovable. Opaque elements still
          // recurse, so a title nested through them (e.g. inside
          // VisuallyHidden) counts.
          isOpaqueElement: (childElement) => {
            const childName = childElement.openingElement.name;
            if (resolveRadixPartName(childName, surface, context) !== null) return false;
            const trailingSegment = getTrailingJsxNameSegment(childName);
            return (
              trailingSegment !== null &&
              /^[A-Z]/.test(trailingSegment) &&
              trailingSegment !== "Fragment"
            );
          },
        });
        if (scan.foundPart || scan.sawOpaqueContent) return;
        context.report({
          node: node.name,
          message: `This ${surface.namespaceName}.Content renders no ${surface.namespaceName}.Title, so the dialog has no accessible name and Radix logs an accessibility error at runtime. Add a Title part (wrapped in VisuallyHidden if the design shows no heading) or an aria-label.`,
        });
        return;
      }
    },
  }),
});

File: native/oxlint/rules/radix-dialog-content-requires-title.rs
use oxc_ast::{ast::JSXAttributeItem, AstKind};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{context::LintContext, rule::Rule, utils::has_jsx_prop_ignore_case, AstNode};

const DIALOG_MESSAGE: &str = "This Dialog.Content renders no Dialog.Title, so the dialog has no accessible name and Radix logs an accessibility error at runtime. Add a Title part (wrapped in VisuallyHidden if the design shows no heading) or an aria-label.";
const ALERT_DIALOG_MESSAGE: &str = "This AlertDialog.Content renders no AlertDialog.Title, so the dialog has no accessible name and Radix logs an accessibility error at runtime. Add a Title part (wrapped in VisuallyHidden if the design shows no heading) or an aria-label.";
const NAME_PROVIDING_ATTRIBUTES: [&str; 3] = ["aria-label", "aria-labelledby", "title"];

#[derive(Debug, Default, Clone)]
pub struct RadixDialogContentRequiresTitle;

declare_oxc_lint!(
    /// Require accessible names on Radix dialog content.
    RadixDialogContentRequiresTitle,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require accessible names on Radix dialog content.",
);

impl Rule for RadixDialogContentRequiresTitle {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if !has_capability_or_unspecified(ctx, "radix-ui") {
            return;
        }
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        for surface_name in ["Dialog", "AlertDialog"] {
            if resolve_radix_dialog_part_name(&opening_element.name, surface_name, ctx)
                != Some("Content")
            {
                continue;
            }
            if opening_element
                .attributes
                .iter()
                .any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)))
                || NAME_PROVIDING_ATTRIBUTES.iter().any(|attribute_name| {
                    has_jsx_prop_ignore_case(opening_element, attribute_name).is_some()
                })
            {
                return;
            }
            let parent = ctx.nodes().parent_node(node.id());
            let AstKind::JSXElement(element) = parent.kind() else {
                return;
            };
            if element.children.is_empty() {
                return;
            }
            let scan = scan_static_jsx_subtree_for_part(
                &element.children,
                ctx,
                |child_name| is_radix_dialog_title_name(child_name, surface_name, ctx),
                |child_element| {
                    let child_name = &child_element.opening_element.name;
                    resolve_radix_dialog_part_name(child_name, surface_name, ctx).is_none()
                        && jsx_element_name_trailing_segment(child_name).is_some_and(|name| {
                            name != "Fragment"
                                && name.as_bytes().first().is_some_and(u8::is_ascii_uppercase)
                        })
                },
            );
            if scan.found_part || scan.saw_opaque_content {
                return;
            }
            let message = if surface_name == "Dialog" {
                DIALOG_MESSAGE
            } else {
                ALERT_DIALOG_MESSAGE
            };
            ctx.diagnostic(OxcDiagnostic::warn(message).with_label(opening_element.name.span()));
            return;
        }
    }
}

fn resolve_radix_dialog_part_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    surface_name: &str,
    ctx: &LintContext<'a>,
) -> Option<&'static str> {
    let primitive_module = match surface_name {
        "Dialog" => "@radix-ui/react-dialog",
        "AlertDialog" => "@radix-ui/react-alert-dialog",
        _ => return None,
    };
    if let Some(api_path) = resolve_jsx_import_api_path(
        element_name,
        |module_source| module_source == primitive_module,
        ctx,
    ) {
        let [part_name] = api_path.as_slice() else {
            return None;
        };
        return radix_dialog_part(part_name);
    }
    let api_path = resolve_jsx_import_api_path(
        element_name,
        |module_source| module_source == "radix-ui",
        ctx,
    )?;
    let [namespace, part_name] = api_path.as_slice() else {
        return None;
    };
    if namespace != surface_name {
        return None;
    }
    radix_dialog_part(part_name)
}

fn radix_dialog_part(part_name: &str) -> Option<&'static str> {
    match part_name {
        "Content" => Some("Content"),
        "Title" => Some("Title"),
        _ => Some("Part"),
    }
}

fn is_radix_dialog_title_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    surface_name: &str,
    ctx: &LintContext<'a>,
) -> bool {
    resolve_radix_dialog_part_name(element_name, surface_name, ctx) == Some("Title")
        || jsx_element_name_trailing_segment(element_name).is_some_and(|name| {
            name == "Title"
                || (surface_name == "Dialog" && name == "DialogTitle")
                || (surface_name == "AlertDialog" && name == "AlertDialogTitle")
        })
}
